### PR TITLE
 Allow setting Azure cloud provider from Kubernetes secrets instread of local configure file

### LINKED ISSUE #27: [k8s] configurable provider
let's get rid of azure.json and use configmaps instead

### PR DESCRIPTION
**Pull request by feiskyer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Refer https://github.com/Azure/container-compute-upstream/issues/27. This PR allows configure Azure cloud provider from Kubernetes secrets.

To support this, a new option  `cloudConfigType` is added. Supported values are:

* file: The cloud provider configuration is read from cloud-config file.
* secret: the cloud provider configuration must be overridden by a secret.
* merge: the cloud provider configuration can be optionally overridden by a secret when it is set explicitly in the secret, this is default value.

Note that the secret is a serialized version of azure.json file with key `cloud-config`. And the secret name is `azure-cloud-provider`.

Since Azure cloud provider would read Kubernetes secrets, the following RBAC should also be configured:

```yaml
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  labels:
    kubernetes.io/cluster-service: "true"
  name: system:azure-cloud-provider-secret-getter
rules:
- apiGroups: [""]
  resources: ["secrets"]
  resourceNames: ["azure-cloud-provider"]
  verbs:
  - get
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  labels:
    kubernetes.io/cluster-service: "true"
  name: system:azure-cloud-provider-secret-getter
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system:azure-cloud-provider-secret-getter
subjects:
- kind: ServiceAccount
  name: azure-cloud-provider
  namespace: kube-system
```

**Which issue(s) this PR fixes**:

Fixes https://github.com/Azure/container-compute-upstream/issues/27

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Azure cloud provider could now be configured by Kubernetes secrets and a new option `cloudConfigType` is introduced, whose candicate values are `file`, `secret` and `merge` (default is `merge`).

action required:

Since Azure cloud provider would read Kubernetes secrets, the following RBAC should be configured:

  apiVersion: rbac.authorization.k8s.io/v1beta1
  kind: ClusterRole
  metadata:
    labels:
      kubernetes.io/cluster-service: "true"
    name: system:azure-cloud-provider-secret-getter
  rules:
  - apiGroups: [""]
    resources: ["secrets"]
    resourceNames: ["azure-cloud-provider"]
    verbs:
    - get
  ---
  apiVersion: rbac.authorization.k8s.io/v1beta1
  kind: ClusterRoleBinding
  metadata:
    labels:
      kubernetes.io/cluster-service: "true"
    name: system:azure-cloud-provider-secret-getter
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: ClusterRole
    name: system:azure-cloud-provider-secret-getter
  subjects:
  - kind: ServiceAccount
    name: azure-cloud-provider
    namespace: kube-system
```

/sig azure
/kind feature
/priority important-soon

/assign @khenidak @andyzhangx

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: 2019

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: I find `secretOverrideTypeMust` is only used in test code, looks like `secretOverrideTypeMust` is a useless option in your code, only for doc

**Review by andyzhangx — COMMENTED** [review]
one question about the code logic, we have two path
```
init --> NewCloud --> initializeCloudFromConfig(fromSecret: false)
Cloud.Initialize --> initializeCloudFromSecret --> initializeCloudFromConfig(fromSecret: true)
```

so `init` would be invoked first and `Cloud.Initialize` could be invoked second, and finally it will goto `initializeCloudFromConfig`, right?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: hmm, it should be checked somewhere. Would add that and ensure only allowed options are accepted.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: good catch

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: Fixed now.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: there are 3 conditions, can we make the logic a little more clear here?
```
if !runingAsKubelet && fromSecret
if !runingAsKubelet && !fromSecret && az.Config.OverrideType != secretOverrideTypeNo
if runingAsKubelet && !config.UseInstanceMetadata
```

e.g.
```
if runingAsKubelet {
 ...
} else {
 ...
}

```

**Review by andyzhangx — COMMENTED** [review]
one question about the code logic, we have two path
```
init --> NewCloud --> initializeCloudFromConfig(fromSecret: false)
Cloud.Initialize --> initializeCloudFromSecret --> initializeCloudFromConfig(fromSecret: true)
```

so `init` would be invoked first and `Cloud.Initialize` could be invoked second, and finally it will goto `initializeCloudFromConfig`, right?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: there're actually four conditions here: runingAsKubelet, fromSecret, UseInstanceMetadata and OverrideType. In that way, there would be a lot embeded if-conditions. Let me think how to make this more easier to read.

Comment (andyzhangx) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: Is this easier to read?
```
if runingAsKubelet {
    if !config.UseInstanceMetadata { ...}
} else {
   if fromSecret {
   ...
   } else {
      if az.Config.OverrideType != secretOverrideTypeNo { ...}
  }
}
```

**Comment by feiskyer** [comment]
> so init would be invoked first and Cloud.Initialize could be invoked second, and finally it will goto initializeCloudFromConfig, right?

Both would go to initializeCloudFromConfig, the difference is Kubelet won't invoke `Initialize`, while controller-manager would do. And for controller-manager, we want `NewCloud` won't error out, and initialize from secret again. But when secret initializing failed, controller-manager would fail with an error.

**Comment by andyzhangx** [comment]
> > so init would be invoked first and Cloud.Initialize could be invoked second, and finally it will goto initializeCloudFromConfig, right?
> 
> Both would go to initializeCloudFromConfig, the difference is Kubelet won't invoke `Initialize`, while controller-manager would do. And for controller-manager, we want `NewCloud` won't error out, and initialize from secret again. But when secret initializing failed, controller-manager would fail with an error.

There are other drivers like CSI driver would also leverage this azure cloud provider lib, it uses `NewCloud`: https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/pkg/azuredisk/azure.go#L45
can we add a switch in `NewCloud`, let it read from secret also? by default, `fromSecret` could be false in `NewCloud`

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: simplified the logic, PTAL

**Comment by feiskyer** [comment]
> can we add a switch in NewCloud, let it read from secret also? by default, fromSecret could be false in NewCloud

`NewCloud()` is a standard interface for cloud provider, hence its interface couldn't be changed. We may export `initializeCloudFromSecret()` for such use cases.

**Comment by andyzhangx** [comment]
> > can we add a switch in NewCloud, let it read from secret also? by default, fromSecret could be false in NewCloud
> 
> `NewCloud()` is a standard interface for cloud provider, hence its interface couldn't be changed. We may export `initializeCloudFromSecret()` for such use cases.

So can you make  `initializeCloudFromSecret()` as  `InitializeCloudFromSecret()`, make it as a public function, thanks.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: can you also make this function as public, that would be more useful as a lib, other part LGTM.

Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: done.

**Review by andyzhangx — APPROVED** [review]
code LGTM
@khenidak  any comments?

**Comment by justaugustus** [comment]
/test pull-kubernetes-e2e-aks-engine-azure
/test pull-kubernetes-integration

/hold
for KEP and enhancement issue
ref: https://github.com/Azure/container-compute-upstream/issues/27#issuecomment-495963963
> > Here is a proposed plan, I am not sure if we need to have this as KEP. if so i can move it to KEP
> 
> @khenidak -- Yes, this will need to be a KEP and an enhancement tracking issue in k/enhancements. Let's get one in place as soon as we can, so we can start to discuss [deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior) of the azure.json config.
> 
> cc: @andrewsykim

/cc @andrewsykim

**Comment by feiskyer** [comment]
pull-kubernetes-e2e-aks-engine-azure is failing because of https://github.com/Azure/aks-engine/issues/1368.

**Review by andyzhangx — APPROVED** [review]
/lgtm
I did a manual verification on this PR, it works well on latest v1.15.0 with `--allow-privileged` parameter removed in the config

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (justaugustus) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]:
```suggestion
	cloudConfigNamespace      = "kube-system"
```

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (justaugustus) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]:
```suggestion
	cloudConfigKey = "cloud-config"
```

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (justaugustus) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]:
```suggestion
type cloudConfigType string
```

**Review by justaugustus — CHANGES_REQUESTED** [review]
Let's reframe this...

If secrets-based configuration is to eventually become the default pattern, then it's more correct to not name its' configuration variables `overrideFoo`.

Additionally, if for whatever reason, we move away from secrets as the configuration mechanism, it'll be easier to refactor if we don't name the variables `secretFoo`.

I've made a few suggestions on naming.

For `cloudConfigType` (added in code suggestions), the constants would then become:

```
const (
	cloudConfigTypeSecret  cloudConfigType = "secret"
	cloudConfigTypeFile    cloudConfigType = "file"
)
```

We should then add a scope of the configuration, which would replace `secretConfigureType`:

```
type cloudConfigScope string

const (
	cloudConfigScopeAll           cloudConfigScope = "all"
	cloudConfigScopeControlPlane  cloudConfigScope = "control-plane"
	cloudConfigScopeNode          cloudConfigScope = "node"
)
```

Let me know what you think!

**Comment by justaugustus** [comment]
In thinking about this a little, I don't think we need a KEP for this. 
This is an implementation detail of an existing enhancement (in-tree Azure cloud provider) and merging it now gives us more time to vet the feature.

Let's make sure to document the impact of this thoroughly (in [docs](https://github.com/kubernetes/cloud-provider-azure/blob/master/docs/cloud-provider-config.md), not just within the code) AND discuss planning to start the deprecation clock for file-based configuration in 1.16.

I'm keeping the original hold on until we have a chance to discuss my review.

**Comment by feiskyer** [comment]
> Let's make sure to document the impact of this thoroughly (in docs, not just within the code) AND discuss planning to start the deprecation clock for file-based configuration in 1.16.

Sure.

**Comment by justaugustus** [comment]
What I had in mind was:
- `cloudConfigTypeSecret` --> all configuration parameters are set from secret
- `cloudConfigTypeFile` --> all configuration parameters are set from file

Maybe there would need to be a third option like:
- `cloudConfigTypeMerge` --> configuration parameters are initially set from file, but overridden if the same parameter is set in the secret

**Comment by feiskyer** [comment]
@justaugustus Addressed comments. PTAL

**Comment by justaugustus** [comment]
Thanks @feiskyer!
/lgtm
/approve

File: staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go]: What's the reason for omitempty changes?

Comment (andrewsykim) [reply staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go]: (makes sense to me, just wondering why)

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: what about hyperkube?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: cloud -> could

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go
Comment (andrewsykim) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: this seems hacky in general though, wondering if there's a better way to do this

**Comment by andrewsykim** [comment]
/approve

Approving for go module changes in staging, will defer to Azure folks for everything else :)

File: staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go]: it's because of merging configs from secret. Without omitempty, all fileds would get an empty value if it's not set. With omitempty, this won't happen.

Comment (andrewsykim) [reply staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go]: Is this rollback safe?

Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go]: yep, there're default values being set after parsing the config.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: Yep, this is a little hacky, any suggestions on this?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config_test.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_config_test.go]: t.Run(test.name, func(t *testing.T) {
    ....
})

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go
Comment (andrewsykim) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: Not sure yet, but probably warrants a follow-up.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: Might be missing something, how is this a merge?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: running

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: I don't agree with these scope values, seems backwards to what we are trying to push for. Secrets should only be available for control plane, if we let clusters define secrets for nodes it make migration harder in the future.

**Review by andrewsykim — COMMENTED** [review]
Some more comments, will defer to Azure maintainers if they are blocking or not :)

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: Same comment as earlier, a kubelet check in the cloud provider seems awkward to me and backwards to out-of-tree patterns.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: Is this Fatal intentional? Seems off?

**Comment by andrewsykim** [comment]
/hold 

So sorry for the last minute hold, but I think the kubelet checks are worth raising a flag for. Can we do this change without the kubelet checks?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: yep, this is a fatal log, because cloud provider won't work without credentials

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_wrap.go]: Yep, let me address this in a separate PR.

**Comment by feiskyer** [comment]
@andrewsykim thanks for reviewing. Agreed, the Kubelet check should be changed in alternative ways. will address these comments before v1.15 release

**Comment by andrewsykim** [comment]
/hold cancel

Thanks @feiskyer

**Comment by andrewsykim** [comment]
heads up @claurence, we might have some follow-up PRs come in for this one

**Comment by andrewsykim** [comment]
/hold

@feiskyer addressing comments given the code freeze extension

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: if some fields are already set in configure file, but not in secret, then it won't be override by another unmarshal. But if it's set in secret, then the value would be override.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure.go]: won't remove this requirement.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go]: yep, would remove this in this PR.

**Comment by feiskyer** [comment]
@andrewsykim Addressed comments. PTAL

/hold cancel

**Review by andrewsykim — COMMENTED** [review]
lgtm, thanks for iterating @feiskyer!

**Comment by jackfrancis** [comment]
@feiskyer will these capabilities ever be backported into pre-1.15 k8s or is this only applicable for 1.15 and greater?

**Comment by andyzhangx** [comment]
> @feiskyer will these capabilities ever be backported into pre-1.15 k8s or is this only applicable for 1.15 and greater?

It won't be backported, only applicable for 1.15 and greater

**Comment by feiskyer** [comment]
@jackfrancis Yep, the feature is only available for v1.15 and greater

**Pull request merged**
